File: src/data/authService.ts
import { DeviceInfo, getMyDeviceProfile } from './syncService';

export interface AuthUser {
  id: string;
  email: string;
  name: string;
  role: 'husband' | 'wife';
  vaultCode?: string;
}

export interface AuthResponse {
  success: boolean;
  token?: string;
  user?: AuthUser;
  vault?: any;
  error?: string;
}

const STORAGE_KEY_TOKEN = 'housevault_auth_token';
const STORAGE_KEY_USER = 'housevault_auth_user';

export function getStoredAuthToken(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem(STORAGE_KEY_TOKEN);
}

export function getStoredUser(): AuthUser | null {
  if (typeof window === 'undefined') return null;
  const raw = localStorage.getItem(STORAGE_KEY_USER);
  try {
    return raw ? JSON.parse(raw) : null;
  } catch {
    return null;
  }
}

export function setStoredAuth(token: string | null, user: AuthUser | null) {
  if (typeof window === 'undefined') return;
  if (token) {
    localStorage.setItem(STORAGE_KEY_TOKEN, token);
  } else {
    localStorage.removeItem(STORAGE_KEY_TOKEN);
  }

  if (user) {
    localStorage.setItem(STORAGE_KEY_USER, JSON.stringify(user));
  } else {
    localStorage.removeItem(STORAGE_KEY_USER);
  }
}

export async function loginAccount(
  email: string,
  password: string
): Promise<AuthResponse> {
  try {
    const device = getMyDeviceProfile(email.includes('haytham') ? 'Haytham' : 'Cati');
    const res = await fetch('/api/auth/login', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password, device })
    });
    const data: AuthResponse = await res.json();
    if (data.success && data.token && data.user) {
      setStoredAuth(data.token, data.user);
    }
    return data;
  } catch (err: any) {
    return { success: false, error: err?.message || 'Eroare de conexiune la server' };
  }
}

export async function registerAccount(
  email: string,
  password: string,
  name: string,
  role: 'husband' | 'wife' = 'husband',
  vaultCode?: string
): Promise<AuthResponse> {
  try {
    const device = getMyDeviceProfile(name);
    const res = await fetch('/api/auth/register', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password, name, role, vaultCode, device })
    });
    const data: AuthResponse = await res.json();
    if (data.success && data.token && data.user) {
      setStoredAuth(data.token, data.user);
    }
    return data;
  } catch (err: any) {
    return { success: false, error: err?.message || 'Eroare de conexiune la server' };
  }
}

export async function checkAuthMe(): Promise<AuthResponse> {
  const token = getStoredAuthToken();
  if (!token) return { success: false, error: 'No token' };

  try {
    const res = await fetch('/api/auth/me', {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });
    const data: AuthResponse = await res.json();
    if (data.success && data.user) {
      setStoredAuth(token, data.user);
    } else {
      setStoredAuth(null, null);
    }
    return data;
  } catch (err: any) {
    return { success: false, error: err?.message };
  }
}

export async function linkVaultAccount(vaultCode: string): Promise<{ success: boolean; vault?: any; error?: string }> {
  const token = getStoredAuthToken();
  if (!token) return { success: false, error: 'Neautorizat' };

  try {
    const device = getMyDeviceProfile();
    const res = await fetch('/api/auth/link-vault', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify({ vaultCode, device })
    });
    return await res.json();
  } catch (err: any) {
    return { success: false, error: err?.message };
  }
}

export function logoutAccount() {
  setStoredAuth(null, null);
}

File: app/src/main/java/com/example/data/AuthService.kt
package com.example.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

data class AuthResult(
    val success: Boolean,
    val token: String? = null,
    val user: AuthUser? = null,
    val vaultData: JSONObject? = null,
    val errorMessage: String? = null
)

object AuthService {

    suspend fun register(
        serverUrl: String,
        email: String,
        password: String,
        name: String,
        role: String = "husband",
        vaultCode: String? = null,
        deviceJson: JSONObject? = null
    ): AuthResult = withContext(Dispatchers.IO) {
        try {
            val cleanUrl = serverUrl.trimEnd('/')
            val endpoint = "$cleanUrl/api/auth/register"
            val url = URL(endpoint)
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
            conn.connectTimeout = 8000
            conn.readTimeout = 8000
            conn.doOutput = true
            conn.doInput = true

            val body = JSONObject().apply {
                put("email", email.trim())
                put("password", password)
                put("name", name.trim())
                put("role", role)
                if (!vaultCode.isNullOrBlank()) put("vaultCode", vaultCode.trim().uppercase())
                if (deviceJson != null) put("device", deviceJson)
            }

            OutputStreamWriter(conn.outputStream, "UTF-8").use { writer ->
                writer.write(body.toString())
                writer.flush()
            }

            val responseCode = conn.responseCode
            val isSuccess = responseCode in 200..299
            val stream = if (isSuccess) conn.inputStream else conn.errorStream
            val responseText = BufferedReader(InputStreamReader(stream, "UTF-8")).use { it.readText() }
            val json = JSONObject(responseText)

            if (isSuccess && json.optBoolean("success")) {
                val token = json.optString("token")
                val userObj = json.optJSONObject("user")
                val user = if (userObj != null) {
                    AuthUser(
                        id = userObj.optString("id"),
                        email = userObj.optString("email"),
                        name = userObj.optString("name"),
                        role = userObj.optString("role", "husband"),
                        vaultCode = userObj.optString("vaultCode").ifBlank { null }
                    )
                } else null
                val vaultData = json.optJSONObject("vault")?.optJSONObject("data")
                AuthResult(success = true, token = token, user = user, vaultData = vaultData)
            } else {
                AuthResult(success = false, errorMessage = json.optString("error", "Eroare la înregistrare"))
            }
        } catch (e: Exception) {
            AuthResult(success = false, errorMessage = e.message ?: "Conexiunea la server a eșuat")
        }
    }

    suspend fun login(
        serverUrl: String,
        email: String,
        password: String,
        deviceJson: JSONObject? = null
    ): AuthResult = withContext(Dispatchers.IO) {
        try {
            val cleanUrl = serverUrl.trimEnd('/')
            val endpoint = "$cleanUrl/api/auth/login"
            val url = URL(endpoint)
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
            conn.connectTimeout = 8000
            conn.readTimeout = 8000
            conn.doOutput = true
            conn.doInput = true

            val body = JSONObject().apply {
                put("email", email.trim())
                put("password", password)
                if (deviceJson != null) put("device", deviceJson)
            }

            OutputStreamWriter(conn.outputStream, "UTF-8").use { writer ->
                writer.write(body.toString())
                writer.flush()
            }

            val responseCode = conn.responseCode
            val isSuccess = responseCode in 200..299
            val stream = if (isSuccess) conn.inputStream else conn.errorStream
            val responseText = BufferedReader(InputStreamReader(stream, "UTF-8")).use { it.readText() }
            val json = JSONObject(responseText)

            if (isSuccess && json.optBoolean("success")) {
                val token = json.optString("token")
                val userObj = json.optJSONObject("user")
                val user = if (userObj != null) {
                    AuthUser(
                        id = userObj.optString("id"),
                        email = userObj.optString("email"),
                        name = userObj.optString("name"),
                        role = userObj.optString("role", "husband"),
                        vaultCode = userObj.optString("vaultCode").ifBlank { null }
                    )
                } else null
                val vaultData = json.optJSONObject("vault")?.optJSONObject("data")
                AuthResult(success = true, token = token, user = user, vaultData = vaultData)
            } else {
                AuthResult(success = false, errorMessage = json.optString("error", "Email sau parolă incorectă"))
            }
        } catch (e: Exception) {
            AuthResult(success = false, errorMessage = e.message ?: "Conexiunea la server a eșuat")
        }
    }

    suspend fun getMe(serverUrl: String, token: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val cleanUrl = serverUrl.trimEnd('/')
            val endpoint = "$cleanUrl/api/auth/me"
            val url = URL(endpoint)
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "GET"
            conn.setRequestProperty("Authorization", "Bearer $token")
            conn.connectTimeout = 8000
            conn.readTimeout = 8000

            val responseCode = conn.responseCode
            if (responseCode in 200..299) {
                val responseText = BufferedReader(InputStreamReader(conn.inputStream, "UTF-8")).use { it.readText() }
                val json = JSONObject(responseText)
                val userObj = json.optJSONObject("user")
                val user = if (userObj != null) {
                    AuthUser(
                        id = userObj.optString("id"),
                        email = userObj.optString("email"),
                        name = userObj.optString("name"),
                        role = userObj.optString("role", "husband"),
                        vaultCode = userObj.optString("vaultCode").ifBlank { null }
                    )
                } else null
                val vaultData = json.optJSONObject("vault")?.optJSONObject("data")
                AuthResult(success = true, token = token, user = user, vaultData = vaultData)
            } else {
                AuthResult(success = false, errorMessage = "Sesiune invalidă")
            }
        } catch (e: Exception) {
            AuthResult(success = false, errorMessage = e.message)
        }
    }
}
